Clarify Send/Sync requirements

Currently it looks like the allocators aren't Send/Sync. Since they are `&mut`, I think they should be?

Here is what I get:
```
error[E0277]: `std::ptr::NonNull<u8>` cannot be sent between threads safely
   --> wgpu-core/src/hub.rs:801:5
    |
800 |     fn test_internal<T: Send + Sync>(_: T) {}
    |                         ---- required by this bound in `hub::_test_send_sync::test_internal`
801 |     test_internal(global)
    |     ^^^^^^^^^^^^^ `std::ptr::NonNull<u8>` cannot be sent between threads safely
    |
    = help: within `[std::option::Option<gpu_alloc::linear::LinearAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>>]`, the trait `std::marker::Send` is not implemented for `std::ptr::NonNull<u8>`
    = note: required because it appears within the type `std::option::Option<std::ptr::NonNull<u8>>`
    = note: required because it appears within the type `gpu_alloc::linear::Chunk<std::sync::Arc<gfx_backend_empty::memory::Memory>>`
    = note: required because it appears within the type `std::option::Option<gpu_alloc::linear::Chunk<std::sync::Arc<gfx_backend_empty::memory::Memory>>>`
    = note: required because it appears within the type `gpu_alloc::linear::Chunks<std::sync::Arc<gfx_backend_empty::memory::Memory>>`
    = note: required because it appears within the type `gpu_alloc::linear::LinearAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>`
    = note: required because it appears within the type `std::option::Option<gpu_alloc::linear::LinearAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>>`
    = note: required because it appears within the type `[std::option::Option<gpu_alloc::linear::LinearAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>>]`
    = note: required because of the requirements on the impl of `std::marker::Send` for `std::ptr::Unique<[std::option::Option<gpu_alloc::linear::LinearAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>>]>`
    = note: required because it appears within the type `std::boxed::Box<[std::option::Option<gpu_alloc::linear::LinearAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>>]>`
    = note: required because it appears within the type `gpu_alloc::GpuAllocator<std::sync::Arc<gfx_backend_empty::memory::Memory>>`
```
So it basically says `linear::Chunk` is not Send/Sync

--- a/gpu-alloc/src/block.rs
+++ b/gpu-alloc/src/block.rs
@@ -19,6 +19,9 @@ pub struct MemoryBlock<M> {
     pub(crate) flavor: MemoryBlockFlavor,
 }
 
+unsafe impl<M> Sync for MemoryBlock<M> where M: Sync {}
+unsafe impl<M> Send for MemoryBlock<M> where M: Send {}
+
 #[derive(Debug)]
 pub(crate) enum MemoryBlockFlavor {
     Dedicated,

--- a/gpu-alloc/src/buddy.rs
+++ b/gpu-alloc/src/buddy.rs
@@ -21,6 +21,9 @@ pub(crate) struct BuddyBlock<M> {
     pub index: usize,
 }
 
+unsafe impl<M> Sync for BuddyBlock<M> where M: Sync {}
+unsafe impl<M> Send for BuddyBlock<M> where M: Send {}
+
 #[derive(Clone, Copy)]
 enum PairState {
     Exhausted,
@@ -272,6 +275,9 @@ pub(crate) struct BuddyAllocator<M> {
     props: MemoryPropertyFlags,
     atom_mask: u64,
 }
+
+unsafe impl<M> Sync for BuddyAllocator<M> where M: Sync {}
+unsafe impl<M> Send for BuddyAllocator<M> where M: Send {}
 
 impl<M> BuddyAllocator<M>
 where

--- a/gpu-alloc/src/linear.rs
+++ b/gpu-alloc/src/linear.rs
@@ -20,6 +20,9 @@ pub(crate) struct LinearBlock<M> {
     pub chunk: u64,
 }
 
+unsafe impl<M> Sync for LinearBlock<M> where M: Sync {}
+unsafe impl<M> Send for LinearBlock<M> where M: Send {}
+
 #[derive(Debug)]
 struct Chunk<M> {
     memory: M,
@@ -59,6 +62,9 @@ pub(crate) struct LinearAllocator<M> {
     props: MemoryPropertyFlags,
     atom_mask: u64,
 }
+
+unsafe impl<M> Sync for LinearAllocator<M> where M: Sync {}
+unsafe impl<M> Send for LinearAllocator<M> where M: Send {}
 
 impl<M> LinearAllocator<M>
 where

